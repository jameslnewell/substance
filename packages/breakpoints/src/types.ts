import {CSSObject, DefaultTheme, ThemeProps} from 'styled-components';

interface ThemedStyleFunction<Theme = DefaultTheme> {
  (props: ThemeProps<Theme>): CSSObject;
}

type Style<Theme = DefaultTheme> = CSSObject | ThemedStyleFunction<Theme>;

export type BreakpointNameConstraint = string | number;
export type BreakpointMap<BreakpointName extends BreakpointNameConstraint> = {
  [name in BreakpointName]: number;
};

export type ResponsiveValueConstraint = string | number | boolean;
export type ResponsiveValueMap<BreakpointName extends BreakpointNameConstraint, Value extends ResponsiveValueConstraint> = {[name in BreakpointName]?: Value};
export type ResponsiveValueOrMap<BreakpointName extends BreakpointNameConstraint, Value extends ResponsiveValueConstraint> = Value | ResponsiveValueMap<BreakpointName, Value>; 

export interface BreakpointFunction<BreakpointName extends BreakpointNameConstraint, Theme = DefaultTheme> {
  (name: BreakpointName): (style: CSSObject) => Style<Theme>;
}

export interface MapFunction<BreakpointName extends BreakpointNameConstraint, Theme = DefaultTheme> {
  <Value extends ResponsiveValueConstraint>(values: Value | ResponsiveValueMap<BreakpointName, Value>, style: (value: Value) => Style<Theme>): Style<Theme>;
}
