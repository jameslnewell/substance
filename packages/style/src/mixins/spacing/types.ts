import {
  MediaNameConstraint,
  MapValueOrMap,
  Style,
  DefaultTheme,
} from '../../types';

export type SpacingNameConstraint = string | number;
export type SpacingMap<SpacingName extends SpacingNameConstraint> = {
  [name in SpacingName]: number | string;
};

export interface SpacingFunction<
  MediaName extends MediaNameConstraint,
  SpacingName extends SpacingNameConstraint,
  Props,
  Theme = DefaultTheme
> {
  (value: MapValueOrMap<MediaName, SpacingName>): Style<Props, Theme>;
}
