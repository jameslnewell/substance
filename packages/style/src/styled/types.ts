/**
 * This file abstracts and adapts the CSS-in-JS framework
 * TODO: create a version of this file for emotion and generate a bundle for emotion
 */
import * as StyledComponents from 'styled-components';
export type Theme = StyledComponents.DefaultTheme;
export type ThemeProps = StyledComponents.ThemeProps<Theme>;
export type CSSFunction = StyledComponents.ThemedCssFunction<
  StyledComponents.DefaultTheme
>;
export type CSSObject = StyledComponents.CSSObject;
export type Interpolation<Props> =
  | StyledComponents.SimpleInterpolation
  | StyledComponents.Interpolation<Props>;
