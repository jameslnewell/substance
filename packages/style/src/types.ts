import {Properties, Pseudos} from 'csstype';

// ========== PROPS ==========

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export type PropsConstraint = {};

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface DefaultProps {}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type ThemeConstraint = any;

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface DefaultTheme {}

export type PropsWithTheme<
  Props extends PropsConstraint,
  Theme extends ThemeConstraint
> = Props & {
  theme: Theme;
};

// ========== STYLE ==========

type CSSProperties = Properties<string | number>;
type CSSPeudoProperties = {[property in Pseudos]?: StyleObject};

export type StyleProperty = keyof CSSProperties;
export type StyleValue<
  Property extends StyleProperty = StyleProperty
> = CSSProperties[Property];
export interface StyleObject extends CSSProperties, CSSPeudoProperties {
  [key: string]: StyleObject | string | number | undefined;
}

export type StyleFunction<Props extends PropsConstraint> = (
  props: Props,
) => Style<Props>;

export type Style<Props extends PropsConstraint> =
  | undefined
  | StyleObject
  | StyleFunction<Props>
  | Array<Style<Props>>;

// ========== FLAT STYLE =========

export type FlatStyleFunction<Props extends PropsConstraint> = (
  props: Props,
) => undefined | StyleObject;

export type FlatStyle<Props extends PropsConstraint> =
  | undefined
  | StyleObject
  | FlatStyleFunction<Props>;

// ========== MEDIA ==========

export type MediaConstraint = string | number;

export type MediaQueries<Media extends MediaConstraint> = {
  [name in Media]: string;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type ValueConstraint = any;
export type ResponsiveValueConstraint = string | number | boolean | undefined;
export type ResponsiveValue<
  Media extends MediaConstraint,
  Value extends ResponsiveValueConstraint
> = Value | {[name in Media]?: Value};

// ========== MATCH MEDIA ==========

export interface MatchFunction<
  Media extends MediaConstraint,
  Theme extends ThemeConstraint = DefaultTheme
> {
  <Props extends PropsConstraint = DefaultProps>(media: Media): (
    style: FlatStyle<PropsWithTheme<Props, Theme>>,
  ) => FlatStyle<PropsWithTheme<Props, Theme>>;
}

// ========== MAP MEDIA ==========

export interface MapFunctionFunction<
  Value extends ResponsiveValueConstraint,
  Theme extends ThemeConstraint = DefaultTheme,
  Props extends PropsConstraint = DefaultProps
> {
  (value: Value): FlatStyle<PropsWithTheme<Props, Theme>>;
}

type MapFunctionReturnValue<
  Theme extends ThemeConstraint = DefaultTheme,
  Props extends PropsConstraint = DefaultProps
> =
  | FlatStyle<PropsWithTheme<Props, Theme>>
  | ((
      props: PropsWithTheme<Props, Theme>,
    ) => FlatStyle<PropsWithTheme<Props, Theme>>[]);

export interface MapFunction<
  Media extends MediaConstraint,
  Theme extends ThemeConstraint = DefaultTheme
> {
  <
    Value extends ResponsiveValueConstraint,
    Props extends PropsConstraint = DefaultProps
  >(
    valueOrValues: ResponsiveValue<Media, Value>,
    fn: MapFunctionFunction<Value, Theme, Props>,
  ): MapFunctionReturnValue<Theme, Props>;
}

// ========== MIXIN ==========

export interface MixinFunction<Value extends ValueConstraint> {
  <Props extends PropsConstraint = DefaultProps>(value: Value): Style<Props>;
}

export interface ResponsiveMixinFunction<
  Media extends MediaConstraint,
  Value extends ResponsiveValueConstraint,
  Theme extends ThemeConstraint = DefaultTheme,
  Props extends PropsConstraint = DefaultProps
> {
  (valueOrValues: ResponsiveValue<Media, Value>): MapFunctionReturnValue<
    Theme,
    Props
  >;
}
