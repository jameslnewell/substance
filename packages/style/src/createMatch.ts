import {
  DefaultTheme,
  MediaConstraint,
  MediaQueries,
  MatchFunction,
  ThemeConstraint,
} from './types';

export interface GetMediaQueriesFromThemeFunction<
  Media extends MediaConstraint,
  Theme extends ThemeConstraint = DefaultTheme
> {
  (theme: Theme | undefined): MediaQueries<Media>;
}

const getQueryFromQueries = <Media extends MediaConstraint>(
  media: Media,
  queries: MediaQueries<Media>,
): string => {
  if (process.env.NODE_ENV !== 'production') {
    if (!Object.prototype.hasOwnProperty.call(queries, media)) {
      console.warn(`Media "${media}" not found in queries.`);
    }
  }
  return queries[media];
};

/**
 * Create a mixin that applies styles when a media query is matched
 *
 * @param options
 *
 * @example
 * const match = createMatch({
 *   mobile: '0em',
 *   tablet: '10em',
 *   desktop: '20em'
 * });
 *
 * const Box = styled.div(
 *   {},
 *   match('mobile')({
 *     color: 'red'
 *   }),
 *   match('tablet')({
 *     color: 'gren'
 *   }),
 *   match('desktop')({
 *     color: 'blue'
 *   })
 * }
 */
export const createMatch = <
  Media extends MediaConstraint,
  Theme extends ThemeConstraint = DefaultTheme
>(
  queries: MediaQueries<Media> | GetMediaQueriesFromThemeFunction<Media, Theme>,
): MatchFunction<Media, Theme> => {
  if (typeof queries === 'function') {
    return (media) => (style) => (props) => {
      return {
        [`@media ${getQueryFromQueries(media, queries(props.theme))}`]: style,
      };
    };
  } else {
    return (media) => (style) => {
      return {
        [`@media ${getQueryFromQueries(media, queries)}`]: style,
      };
    };
  }
};
