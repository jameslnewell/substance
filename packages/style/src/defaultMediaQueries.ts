import {MediaQueries} from './types';

export type DefaultMedia = 'mobile' | 'tablet' | 'desktop';

export const defaultMediaQueries: MediaQueries<DefaultMedia> = {
  mobile: '(min-width: 0em)',
  tablet: '(min-width: 46.0625em)', // 737px
  desktop: '(min-width: 64.0625em)', // 1025px
};
