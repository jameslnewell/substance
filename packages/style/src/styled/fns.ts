/**
 * This file abstracts and adapts the CSS-in-JS framework
 * TODO: create a version of this file for emotion and generate a bundle for emotion
 */
export {default as styled, css} from 'styled-components';
