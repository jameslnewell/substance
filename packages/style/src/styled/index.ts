/**
 * This file abstracts and adapts a specific CSS-in-JS framework
 * TODO: create a version of this file for emotion and generate a bundle for emotion
 */

// ========== TYPES ==========

import * as StyledComponents from 'styled-components';

export type Theme = StyledComponents.DefaultTheme;

export type CSSFunction = StyledComponents.ThemedCssFunction<
  StyledComponents.DefaultTheme
>;
export type CSSObject = StyledComponents.CSSObject;
export type Interpolation<Props> =
  | StyledComponents.SimpleInterpolation
  | StyledComponents.Interpolation<Props>;

// ========== FUNCTIONS ==========

export {default as styled, css} from 'styled-components';
