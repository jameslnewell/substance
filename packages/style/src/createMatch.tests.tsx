import 'jest-styled-components';
import React from 'react';
import styled from 'styled-components';
import {render} from '@testing-library/react';
import {keysOf} from '@substance/test-utilities';
import {MediaQueries, MatchFunction, MediaConstraint} from './types';
import {defaultMediaQueries} from './defaultMediaQueries';
import {createMatch} from './createMatch';

function applyStyleForEachMedia<Media extends MediaConstraint>(
  queries: MediaQueries<Media>,
  match: MatchFunction<Media>,
) {
  return keysOf<MediaQueries<Media>>(queries).map((media) =>
    match(media)({
      content: `"${media}"`,
    }),
  );
}

function expectStyleForEachMedia<Media extends MediaConstraint>(
  queries: MediaQueries<Media>,
  element: ChildNode | null,
) {
  const keys = keysOf(queries);
  expect.assertions(keys.length * 2); // FIXME: bit confused by this!
  keys.forEach((media) => {
    expect(element).toHaveStyleRule('content', `"${media}"`, {
      media: queries[media],
    });
  });
}

describe('createMatch()', () => {
  test('no style is applied when media query is matched but style is undefined', () => {
    const match = createMatch(defaultMediaQueries);
    const Component = styled.div({}, match('mobile')(undefined));
    const {container} = render(<Component />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('no style is applied when media query is matched but style is a function returning undefined', () => {
    const match = createMatch(defaultMediaQueries);
    const Component = styled.div(
      {},
      match('mobile')(() => undefined),
    );
    const {container} = render(<Component />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('style is applied when media query is matched and style is an object', () => {
    const match = createMatch(defaultMediaQueries);
    const Component = styled.div({}, match('mobile')({color: 'red'}));
    const {container} = render(<Component />);
    expect(container.firstChild).toHaveStyleRule('color', 'red', {
      media: defaultMediaQueries['mobile'],
    });
  });

  test('style is applied when media query is matched and style is a function returning an object', () => {
    const match = createMatch(defaultMediaQueries);
    const Component = styled.div(
      {},
      match('mobile')(() => ({color: 'red'})),
    );
    const {container} = render(<Component />);
    expect(container.firstChild).toHaveStyleRule('color', 'red', {
      media: defaultMediaQueries['mobile'],
    });
  });

  describe('when using a queries object', () => {
    const match = createMatch(defaultMediaQueries);
    const Component = styled.div(
      {},
      applyStyleForEachMedia(defaultMediaQueries, match),
    );

    test('style is applied when media query is matched', () => {
      const {container} = render(<Component />);
      expectStyleForEachMedia(defaultMediaQueries, container.firstChild);
    });
  });

  describe('when using a queries function', () => {
    const match = createMatch(() => defaultMediaQueries);
    const Component = styled.div(
      {},
      applyStyleForEachMedia(defaultMediaQueries, match),
    );

    test('style is applied when media query is matched', () => {
      const {container} = render(<Component />);
      expectStyleForEachMedia(defaultMediaQueries, container.firstChild);
    });
  });
});
