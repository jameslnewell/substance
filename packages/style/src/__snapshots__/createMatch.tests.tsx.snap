// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createMatch() no style is applied when media query is matched but style is a function returning undefined 1`] = `
<div
  class=""
/>
`;

exports[`createMatch() no style is applied when media query is matched but style is undefined 1`] = `
<div
  class=""
/>
`;
